### PR TITLE
Autoscaling capacity off transport thread

### LINKED ISSUE #79104: Autoscaling capacity calculation blocks a transport thread
**Elasticsearch version** (`bin/elasticsearch --version`): seen in 7.13.1 but `master` also affected

**Plugins installed**: [Cloud]

**JVM version** (`java -version`): Bundled

**OS version** (`uname -a` if on a Unix-like system): Cloud

**Description of the problem including expected versus actual behavior**:

In a 100-node cluster with ~23k shards we saw the get-autoscaling-capacity API taking 10-20 seconds to complete its calculations. However this API runs on `SAME`, so while the calculation was ongoing it blocked other network activity on the same worker.

### PR DESCRIPTION
**Pull request by henningandersen**
Calculating the autoscaling capacity can take considerable time and
has therefore been moved off transport threads. Also, we ensure that
only one thread calculates capacity and only compute one result if
multiple requests are pending, serving all pending requests in one go.
Finally, the capacity requests are now cancellable.

Closes #79104

### CODE REVIEW

File: x-pack/plugin/autoscaling/src/internalClusterTest/java/org/elasticsearch/xpack/autoscaling/action/GetAutoscalingCapacityRestCancellationIT.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+                restClient.performRequestAsync(getCapacityRequest, wrapAsRestResponseListener(successFuture2));
+                assertBusy(
+                    () -> assertThat(
+                        internalCluster().getMasterNodeInstance(TransportGetAutoscalingCapacityAction.class).responseCacheQueueSize(),
+                        equalTo(3)
+                    )
+                );
+                cancellable.cancel();
+                waitForCancelledCapacityTask();
+            });
```
Comment (DaveCTurner) [inline x-pack/plugin/autoscaling/src/internalClusterTest/java/org/elasticsearch/xpack/autoscaling/action/GetAutoscalingCapacityRestCancellationIT.java:103]: Can we document that the other tasks are still blocked here?

```suggestion

                assertFalse(successFuture1.isDone());
                assertFalse(successFuture2.isDone());
            });
```

File: x-pack/plugin/autoscaling/src/test/java/org/elasticsearch/xpack/autoscaling/action/CapacityResponseCacheTests.java [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+        PlainActionFuture<Integer> blockingFuture = new PlainActionFuture<>();
+        cache.get(() -> false, blockingFuture);
+        await(barrier);
+
+        int count = between(1, 10);
+        List<PlainActionFuture<Integer>> supersededFutures = new ArrayList<>(count + 1);
+        for (int i = 0; i < count; ++i) {
+            PlainActionFuture<Integer> supersededFuture = new PlainActionFuture<>();
+            supersededFutures.add(supersededFuture);
+            cache.get(() -> false, supersededFuture);
```
Comment (DaveCTurner) [inline x-pack/plugin/autoscaling/src/test/java/org/elasticsearch/xpack/autoscaling/action/CapacityResponseCacheTests.java:107]: Could we call this from various threads, not just sequentially?

File: x-pack/plugin/autoscaling/src/test/java/org/elasticsearch/xpack/autoscaling/action/CapacityResponseCacheTests.java [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+
+        CyclicBarrier barrier = new CyclicBarrier(2);
+        responses.add(() -> {
+            await(barrier);
+            await(barrier);
+            return 0;
+        });
+
+        PlainActionFuture<Integer> blockingFuture = new PlainActionFuture<>();
+        cache.get(() -> false, blockingFuture);
```
Comment (DaveCTurner) [inline x-pack/plugin/autoscaling/src/test/java/org/elasticsearch/xpack/autoscaling/action/CapacityResponseCacheTests.java:99]: Could we fire off multiple calls (concurrently) all of which are expected to block (at least one of which will return 0, although some of them might return `response`)

File: x-pack/plugin/autoscaling/src/test/java/org/elasticsearch/xpack/autoscaling/action/CapacityResponseCacheTests.java [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+
+        // release the first blocking thread.
+        await(barrier);
+        assertThat(blockingFuture.actionGet(), equalTo(0));
+
+        await(responseBarrier);
+        cancelled.set(true);
+        await(responseBarrier);
+        supersededFutures.forEach(future -> expectThrows(TaskCancelledException.class, future::actionGet));
+        expectThrows(TaskCancelledException.class, responseFuture::actionGet);
```
Comment (DaveCTurner) [inline x-pack/plugin/autoscaling/src/test/java/org/elasticsearch/xpack/autoscaling/action/CapacityResponseCacheTests.java:185]: Could we also have a test that verifies that if a strict subset of the jobs are cancelled then execution still proceeds?

**Review by DaveCTurner — COMMENTED** [review]
Looks good - I left a couple of nits and a few test suggestions.

File: x-pack/plugin/autoscaling/src/internalClusterTest/java/org/elasticsearch/xpack/autoscaling/action/GetAutoscalingCapacityRestCancellationIT.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+    @Override
+    protected boolean ignoreExternalCluster() {
+        return true;
+    }
+
+    public void testCapacityRestCancellationAndResponse() throws Exception {
+        internalCluster().startMasterOnlyNode();
+
+        putAutoscalingPolicy(Map.of("count", Settings.EMPTY, "!sync", Settings.builder().put("check_for_cancel", randomBoolean()).build()));
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply x-pack/plugin/autoscaling/src/internalClusterTest/java/org/elasticsearch/xpack/autoscaling/action/GetAutoscalingCapacityRestCancellationIT.java]: Sure, see cf9d5d4013f8eb4ff9a03990a7144669ba4c2fd4

File: x-pack/plugin/autoscaling/src/internalClusterTest/java/org/elasticsearch/xpack/autoscaling/AutoscalingSyncTestDeciderService.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            assert false;
+        }
+        return null;
+    }
+
+    private void internalSync() {
+        try {
+            syncBarrier.await();
+        } catch (InterruptedException | BrokenBarrierException e) {
+            assert false;
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply x-pack/plugin/autoscaling/src/internalClusterTest/java/org/elasticsearch/xpack/autoscaling/AutoscalingSyncTestDeciderService.java]: :thu: cf9d5d4013f8eb4ff9a03990a7144669ba4c2fd4

File: x-pack/plugin/autoscaling/src/internalClusterTest/java/org/elasticsearch/xpack/autoscaling/action/GetAutoscalingCapacityRestCancellationIT.java (hunk 1 above)
Comment (henningandersen) [reply x-pack/plugin/autoscaling/src/internalClusterTest/java/org/elasticsearch/xpack/autoscaling/action/GetAutoscalingCapacityRestCancellationIT.java:115]: :thu: cf9d5d4013f8eb4ff9a03990a7144669ba4c2fd4

File: x-pack/plugin/autoscaling/src/test/java/org/elasticsearch/xpack/autoscaling/action/CapacityResponseCacheTests.java (hunk 2 above)
Comment (henningandersen) [reply x-pack/plugin/autoscaling/src/test/java/org/elasticsearch/xpack/autoscaling/action/CapacityResponseCacheTests.java]: Absolutely, done in ed38bcb31f681d0297dec34f7b0bf9cd36a3c476, though your next comment revised some of the code changes in ed38bcb31f681d0297dec34f7b0bf9cd36a3c476

File: x-pack/plugin/autoscaling/src/test/java/org/elasticsearch/xpack/autoscaling/action/CapacityResponseCacheTests.java (hunk 3 above)
Comment (henningandersen) [reply x-pack/plugin/autoscaling/src/test/java/org/elasticsearch/xpack/autoscaling/action/CapacityResponseCacheTests.java]: Good idea, done in ed38bcb31f681d0297dec34f7b0bf9cd36a3c476

File: x-pack/plugin/autoscaling/src/test/java/org/elasticsearch/xpack/autoscaling/action/CapacityResponseCacheTests.java (hunk 4 above)
Comment (henningandersen) [reply x-pack/plugin/autoscaling/src/test/java/org/elasticsearch/xpack/autoscaling/action/CapacityResponseCacheTests.java:205]: I added this in the prior commit/test by [randomly letting the superseded tasks cancel](https://github.com/elastic/elasticsearch/pull/80499/commits/ed38bcb31f681d0297dec34f7b0bf9cd36a3c476#diff-84ea97ed99c61e1fcb20519981e9ec441f105817d74211ae7b6db06fe3387d61R127), I think that is good enough?

Comment (DaveCTurner) [reply x-pack/plugin/autoscaling/src/test/java/org/elasticsearch/xpack/autoscaling/action/CapacityResponseCacheTests.java:205]: Yep looks good :)

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by henningandersen** [comment]
Thanks David!

**Comment by henningandersen** [comment]
7.16 backport is pending.

**Pull request merged**